File: 100_prisoners.py
# https://rosettacode.org/wiki/100_prisoners
import random
from typing import Callable

def run_checks(f):
    return f.__code__.co_argcount == 2


def create_prisoners():
    prisoners = list(range(1, 101))
    boxes = dict(zip(random.sample(range(1, 101), 100), range(1, 101)))
    return prisoners, boxes

def naive_strategy(prisoners: list[int], boxes: dict[int, int]):
    for prisoner in prisoners:
        for i in range(50):
            box_opened = random.randint(1, 100)
            if boxes[box_opened] == prisoner:
                break
        else:
            return False
    return True

def smart_strategy(prisoners: list[int], boxes: dict[int, int]):
    for prisoner in prisoners:
        box_opened = prisoner
        for i in range(50):
            if (b := boxes[box_opened]) == prisoner:
                break
            box_opened = b
        else:
            return False
    return True

def run_strategy(strategy: Callable, iterations: int = 10_000):
    if not run_checks(strategy):
        return ValueError('Invalid function')

    success = 0
    for i in range(iterations):
        success += int(strategy(*create_prisoners()))
    return (success / iterations) * 100

def main(_round=2):
    print(f"Naive strategy success rate: {run_strategy(naive_strategy):.{_round}f}%")
    print(f"Smart strategy success rate: {run_strategy(smart_strategy):.{_round}f}%")

if __name__ == '__main__':
    main()

File: nim/100_prisoners.nim
import std/random
import std/sequtils
import std/strformat
import tables

proc make_boxes(): Table[int, int] =
  var
    boxes = initTable[int, int]()
    keys = toSeq(0..99)

  shuffle(keys)
  for i in 0..99:
    boxes[i] = keys[i]

  return boxes

proc make_prionsers(): seq[int] =
  var
    prionsers = toSeq(0..99)
  shuffle(prionsers)
  return prionsers

proc naive_strategy(prisoners: seq[int], boxes: Table[int, int]): bool =
  var
    box_opened = 0
    prev = newSeq[int]()
    found: bool = false
  for prisoner in prisoners:
    prev = newSeq[int]()
    for i in 1..50:
      box_opened = rand(0..99)
      while box_opened in prev:
        box_opened = rand(0..99)
      prev.add(box_opened)

      if boxes[box_opened] == prisoner:
        break
    if boxes[box_opened] != prisoner:
      return false
  return true

proc smart_strategy(prisoners: seq[int], boxes: Table[int, int]): bool =
  var
    box_opened: int = 0
    b: int = 0
  for prisoner in prisoners:
    box_opened = prisoner
    for i in 1..50:
      b = boxes[box_opened]
      if b == prisoner:
        break
      box_opened = b
    if b != prisoner:
      return false
  return true

proc run_strategy(strategy: proc, iterations: int = 10_000): float =
  var
    success = 0
    c = false
  for i in 0..iterations:
    var
      prisoners = make_prionsers()
      boxes = make_boxes()

    c = strategy(prisoners, boxes)
    success += (int) c

  return (success / iterations) * 100

proc main() =
  randomize()
  echo fmt"Naive strategy success rate: {run_strategy(naive_strategy)}%"
  echo fmt"Smart strategy success rate: {run_strategy(smart_strategy)}%"

main()

